lightning/src/offers/invoice.rs

```diff
@@ -478,108 +630,60 @@ impl Bolt12Invoice {
-		match absolute_expiry {
-			Some(seconds_from_epoch) => match SystemTime::UNIX_EPOCH.elapsed() {
-				Ok(elapsed) => elapsed > seconds_from_epoch,
-				Err(_) => false,
-			},
-			None => false,
-		}
+	pub fn is_expired(&$self) -> bool {
+		$contents.is_expired()
 	}
 
 	/// SHA256 hash of the payment preimage that will be given in return for paying the invoice.
-	pub fn payment_hash(&self) -> PaymentHash {
-		self.contents.fields().payment_hash
+	pub fn payment_hash(&$self) -> PaymentHash {
+		$contents.payment_hash()
 	}
 
 	/// The minimum amount required for a successful payment of the invoice.
-	pub fn amount_msats(&self) -> u64 {
-		self.contents.fields().amount_msats
+	pub fn amount_msats(&$self) -> u64 {
+		$contents.amount_msats()
 	}
 
 	/// Fallback addresses for paying the invoice on-chain, in order of most-preferred to
 	/// least-preferred.
-	pub fn fallbacks(&self) -> Vec<Address> {
-		let network = match self.network() {
-			None => return Vec::new(),
-			Some(network) => network,
-		};
-
-		let to_valid_address = |address: &FallbackAddress| {
-			let version = match WitnessVersion::try_from(address.version) {
-				Ok(version) => version,
-				Err(_) => return None,
-			};
-
-			let program = &address.program;
-			if program.len() < 2 || program.len() > 40 {
-				return None;
-			}
-
-			let address = Address {
-				payload: Payload::WitnessProgram {
-					version,
-					program: address.program.clone(),
-				},
-				network,
-			};
-
-			if !address.is_standard() && version == WitnessVersion::V0 {
-				return None;
-			}
-
-			Some(address)
-		};
-
-		self.contents.fields().fallbacks
-			.as_ref()
-			.map(|fallbacks| fallbacks.iter().filter_map(to_valid_address).collect())
-			.unwrap_or_else(Vec::new)
-	}
-
-	fn network(&self) -> Option<Network> {
-		let chain = self.contents.chain();
-		if chain == ChainHash::using_genesis_block(Network::Bitcoin) {
-			Some(Network::Bitcoin)
-		} else if chain == ChainHash::using_genesis_block(Network::Testnet) {
-			Some(Network::Testnet)
-		} else if chain == ChainHash::using_genesis_block(Network::Signet) {
-			Some(Network::Signet)
-		} else if chain == ChainHash::using_genesis_block(Network::Regtest) {
-			Some(Network::Regtest)
-		} else {
-			None
-		}
+	pub fn fallbacks(&$self) -> Vec<Address> {
+		$contents.fallbacks()
 	}
 
 	/// Features pertaining to paying an invoice.
-	pub fn features(&self) -> &Bolt12InvoiceFeatures {
-		&self.contents.fields().features
+	pub fn invoice_features(&$self) -> &Bolt12InvoiceFeatures {
+		$contents.features()
 	}
 
 	/// The public key corresponding to the key used to sign the invoice.
-	pub fn signing_pubkey(&self) -> PublicKey {
-		self.contents.fields().signing_pubkey
+	pub fn signing_pubkey(&$self) -> PublicKey {
+		$contents.signing_pubkey()
 	}
+} }
+
+impl UnsignedBolt12Invoice {
+	invoice_accessors!(self, self.contents);
+}
+
+impl Bolt12Invoice {
+	invoice_accessors!(self, self.contents);
```

Comment:
If we added a parameter for `Offer` vs `Refund` to the `{Unsigned}Bolt12Invoice` type signatures, could we use `{offer,refund}_accessors` macros for each respective variant and avoid all the boilerplate-y methods below?

Reply:
Unfortunately not. `Refund` doesn't contain the same fields as `Offer` and includes `InvoiceRequest` fields. There are some `Option` mismatches, too.

Reply:
Hmm... actually let me think a bit more on this... I suspect it would sorta be done but the real problem is overlapping methods from each. I guess a prefix may work, but bindings might be gross.

Reply:
I mean having a parameterized type where one set of methods is defined for one parameterization and another set is defined for another parameterization.